client/received_blocks_map.go

```diff
@@ -0,0 +1,303 @@
+		finalizeKey: !opts.NoFinalizeKey,
+	})
+}
+
+type mapKeyOptions struct {
+	copyKey     bool
+	finalizeKey bool
+}
+
+func (m *recievedBlocksMap) set(k idAndBlockStart, v receivedBlocks, opts mapKeyOptions) {
+	hash := m.hash(k)
+	for entry, ok := m.lookup[hash]; ok; entry, ok = m.lookup[hash] {
+		if m.equals(entry.key.key, k) {
+			m.lookup[hash] = mapEntry{
+				key:   entry.key,
+				value: v,
+			}
+			return
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+
+	m.lookup[hash] = mapEntry{
+		key:   m.newMapKey(k, opts),
+		value: v,
+	}
+}
+
+// Iter provides the underlying map to allow for using a native Go for loop
+// to iterate the map, however callers should only ever read and not write
+// the map.
+func (m *recievedBlocksMap) Iter() map[mapHash]mapEntry {
+	return m.lookup
+}
+
+// Len returns the number of map entries in the map.
+func (m *recievedBlocksMap) Len() int {
+	return len(m.lookup)
+}
+
+// Contains returns true if value exists for key, false otherwise, it is
+// shorthand for a call to Get that doesn't return the value.
+func (m *recievedBlocksMap) Contains(k idAndBlockStart) bool {
+	_, ok := m.Get(k)
+	return ok
+}
+
+// Delete will remove a value set in the map for the specified key.
+func (m *recievedBlocksMap) Delete(k idAndBlockStart) {
+	hash := m.hash(k)
+	for entry, ok := m.lookup[hash]; ok; entry, ok = m.lookup[hash] {
+		if m.equals(entry.key.key, k) {
+			m.removeMapKey(hash, entry.key)
+			return
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+}
+
+// Reset will reset the map by simply deleting all keys to avoid
+// allocating a new map.
+func (m *recievedBlocksMap) Reset() {
+	for hash, entry := range m.lookup {
+		m.removeMapKey(hash, entry.key)
+	}
+}
+
+// Reallocate will avoid deleting all keys and reallocate a new
+// map, this is useful if you believe you have a large map and
+// will not need to grow back to a similar size.
+func (m *recievedBlocksMap) Reallocate() {
+	if m.initialSize > 0 {
+		m.lookup = make(map[mapHash]mapEntry, m.initialSize)
+	} else {
+		m.lookup = make(map[mapHash]mapEntry)
+	}
+}
+
+// id is a small utility type to avoid the heavy-ness of the true ident.ID
+// implementation when copying keys just for the use of looking up keys in
+// the map internally.
+type id []byte
+
+// var declaration to ensure package type id implements ident.ID
+var _ ident.ID = id(nil)
+
+func (v id) Data() checked.Bytes {
+	// Data is not called by the generated hashmap code, hence we don't
+	// implement this and no callers will call this as the generated code
+	// is the only user of this type.
+	panic("not implemented")
+}
+
+func (v id) Bytes() []byte {
+	return v
+}
+
+func (v id) String() string {
+	return string(v)
+}
+
+func (v id) Equal(value ident.ID) bool {
+	return bytes.Equal(value.Bytes(), v)
+}
+
+func (v id) Reset() {
+	for i := range v {
+		v[i] = byte(0)
+	}
+}
+
+func (v id) Finalize() {
```

Comment:
would it be better to force users to only use the non-copy version of `Set`, and then panic inside this method?